zookeeper-server/src/main/java/org/apache/zookeeper/server/persistence/FileTxnLog.java

```diff
@@ -289,22 +297,25 @@ public synchronized boolean append(TxnHeader hdr, Record txn, TxnDigest digest)
             logStream = new BufferedOutputStream(fos);
             oa = BinaryOutputArchive.getArchive(logStream);
             FileHeader fhdr = new FileHeader(TXNLOG_MAGIC, VERSION, dbId);
+            long dataSize = oa.getDataSize();
             fhdr.serialize(oa, "fileheader");
             // Make sure that the magic number is written before padding.
             logStream.flush();
-            filePadding.setCurrentSize(fos.getChannel().position());
+            filePosition += oa.getDataSize() - dataSize;
+            filePadding.setCurrentSize(filePosition);
             streamsToFlush.add(fos);
         }
-        filePadding.padFile(fos.getChannel());
+        fileSize = filePadding.padFile(fos.getChannel(), filePosition);
         byte[] buf = Util.marshallTxnEntry(hdr, txn, digest);
         if (buf == null || buf.length == 0) {
             throw new IOException("Faulty serialization for header " + "and txn");
         }
+        long dataSize = oa.getDataSize();
         Checksum crc = makeChecksumAlgorithm();
         crc.update(buf, 0, buf.length);
         oa.writeLong(crc.getValue(), "txnEntryCRC");
         Util.writeTxnBytes(oa, buf);
-
+        unFlushedSize += oa.getDataSize() - dataSize;
```

Comment:
Can you describe a bit how this calculation is correct?

Reply:
UnFlushedSize: The position of the file which already write to the channel, but not flush to the channel.

In this case, we record the dataSize before writing data. Then write `CRC`, `serializeData`, but we didn't flush. So we use the current dataSize minus the before dataSize, to get the unFlushedSize.

In the commit, we will add the unFlushedSize to the filePosition.

Reply:
please add a comment in the code, it is not straightforward

Reply:
addressed